### ISSUE TITLE
[Bug] Padding tokens incorrectly included in response mask when EOS is absent

### ISSUE THREAD

**Issue by HareshKarnan** [volcengine/verl#3805]
### System Info

🐛 Bug Report

Padding tokens are incorrectly treated as valid tokens in response masks when EOS tokens are absent, causing them to contribute to training loss calculations.

Root Cause:

The get_response_mask() function in verl/utils/torch_functional.py:225-245 only creates masks based on EOS token positions. When no EOS token is present in a response, it returns all ones:

# From verl/utils/torch_functional.py:244-245  
eos_mask = torch.isin(response_id, torch.tensor(eos_token, device=response_id.device)).int()  
return (eos_mask.cumsum(dim=1) - eos_mask).eq(0)
When eos_mask is all zeros (no EOS), cumsum returns all zeros, and (0 - 0).eq(0) returns all True, marking padding tokens as valid.

Affected Code Locations:

verl/workers/rollout/vllm_rollout/vllm_rollout_spmd.py:388-390
verl/workers/rollout/sglang_rollout/sglang_rollout.py:773-775
verl/workers/rollout/hf_rollout.py:156-158
Impact:

Padding tokens erroneously participate in:

KL divergence computation
GAE advantage estimation
Advantage whitening/normalization
GRPO group statistics
Policy loss calculation
This introduces gradient noise and training instability, especially with larger group sizes (e.g., n=8) where multiple samples may lack EOS tokens.

Steps to Reproduce:

Run any PPO/GRPO training with sequences that may not generate EOS tokens (e.g., max_length truncation)
Observe that response_attention_mask includes padding positions when EOS is absent
Verify that these padding tokens contribute to loss calculations
Example Scenario:

# response = [5, 6, 7, 8, 0, 0, 0] where pad_token_id=0, eos_token_id=2  
# No EOS present  
# get_response_mask returns [1, 1, 1, 1, 1, 1, 1]  
# Padding tokens at positions 4-6 incorrectly marked as valid
Expected Behavior
Response masks should always exclude padding tokens, regardless of whether EOS is present. The mask should be computed as:

response_attention_mask = get_response_mask(  
    response_id=response, eos_token=eos_token_id, dtype=attention_mask.dtype  
) * (response != pad_token_id).to(attention_mask.dtype)
This ensures:

Padding tokens are always masked out
Edge case where pad_token_id == eos_token_id is handled correctly
Training stability is improved by eliminating gradient contributions from padding
Additional Context:

This affects all three rollout backends (vLLM, SGLang, HF)
GRPO is particularly vulnerable as it doesn't validate that groups have at least 2 valid completions
The bug has existed since the initial implementation of get_response_mask()

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Run any PPO/GRPO training with sequences that may not generate EOS tokens (e.g., max_length truncation)
Observe that response_attention_mask includes padding positions when EOS is absent
Verify that these padding tokens contribute to loss calculations
Example Scenario:

# response = [5, 6, 7, 8, 0, 0, 0] where pad_token_id=0, eos_token_id=2  
# No EOS present  
# get_response_mask returns [1, 1, 1, 1, 1, 1, 1]  
# Padding tokens at positions 4-6 incorrectly marked as valid

### Expected behavior

Expected Behavior
Response masks should always exclude padding tokens, regardless of whether EOS is present. The mask should be computed as:

response_attention_mask = get_response_mask(  
    response_id=response, eos_token=eos_token_id, dtype=attention_mask.dtype  
) * (response != pad_token_id).to(attention_mask.dtype)
This ensures:

Padding tokens are always masked out
Edge case where pad_token_id == eos_token_id is handled correctly
Training stability is improved by eliminating gradient contributions from padding
Additional Context:

This affects all three rollout backends (vLLM, SGLang, HF)
GRPO is particularly vulnerable as it doesn't validate that groups have at least 2 valid completions
The bug has existed since the initial implementation of get_response_mask()

**Comment by pavelgein** [comment]
Hi @HareshKarnan, we have encountered the same problem, but the root cause (beside this quite strange behaviour) was that vllm max_model_len was set to a value less  than max_prompt_length + max_response_length

Because the code pad to max_response_length and this the only (as far as I see) reason, when the model does not generate eos-token and the response is still padded

**Comment by Kirrito-k423** [comment]
本 issue 根据定时清理规则关闭。主任务已检查该 issue 的 comments 和 timeline：最近一次非当前运行账号（Kirrito-k423）的公开回复或引用为 2025-11-12T11:55:51Z（event=mentioned, actor=HareshKarnan，证据：https://api.github.com/repos/verl-project/verl/issues/events/20886635382），距本轮运行日期 2026-07-21 已超过 6 个月。检查来源：https://api.github.com/repos/verl-project/verl/issues/3805/comments 和 https://api.github.com/repos/verl-project/verl/issues/3805/timeline。如果作者或社区成员仍能提供新的复现、证据或需求背景，欢迎重新评论请求重开，或创建新的 issue。

Closing this issue under the scheduled cleanup rule. The task checked this issue's comments and timeline: the latest public reply or reference from an account other than the current automation account (Kirrito-k423) was at 2025-11-12T11:55:51Z (event=mentioned, actor=HareshKarnan, evidence: https://api.github.com/repos/verl-project/verl/issues/events/20886635382), which is more than 6 months before the run date 2026-07-21. Sources checked: https://api.github.com/repos/verl-project/verl/issues/3805/comments and https://api.github.com/repos/verl-project/verl/issues/3805/timeline. If the author or community can still provide a fresh reproduction, evidence, or requirements context, please comment to request reopening or open a new issue.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
